File: calculator/regimes/patent.py
from __future__ import annotations

from ..constants import THRESHOLD_1_PERCENT
from ..models import CalcInput, CalcResult, CalculationContext

PATENT_RATE = 0.06


def calculate_patent(data: CalcInput, ctx: CalculationContext) -> CalcResult:
    revenue = data.revenue
    expenses_total = ctx.cost_of_goods + data.rent + ctx.other_expenses + ctx.annual_fot

    tax_before_deduction = max(data.patent_cost_year, 0.0)
    manual_pvd = max(data.patent_pvd_period, 0.0)
    auto_pvd = tax_before_deduction / PATENT_RATE if tax_before_deduction > 0 else 0.0
    pvd_used = manual_pvd if manual_pvd > 0 else auto_pvd

    contrib_workers = ctx.insurance_standard
    owner_extra_base = pvd_used
    owner_extra = max(owner_extra_base - THRESHOLD_1_PERCENT, 0.0) * 0.01
    contrib_self = data.fixed_contrib + owner_extra

    deductible_contrib = contrib_self + contrib_workers
    has_employees_limit = ctx.annual_fot > 0.0 and data.employees > 0
    deduction_limit_ratio = 0.5 if has_employees_limit else 1.0
    deduction_limit = tax_before_deduction * deduction_limit_ratio
    tax_deduction = min(deductible_contrib, deduction_limit)
    tax_payable = max(tax_before_deduction - tax_deduction, 0.0)

    net_profit_accounting = revenue - expenses_total - tax_payable
    net_profit_cash = net_profit_accounting - contrib_self - contrib_workers

    total_burden = tax_payable + contrib_self + contrib_workers
    burden_percent = (total_burden / revenue * 100) if revenue > 0 else 0.0

    pvd_source = "input" if manual_pvd > 0 else "auto"

    extra = {
        "income_without_vat": revenue,
        "expenses_without_vat": expenses_total,
        "net_profit_accounting": net_profit_accounting,
        "net_profit_cash": net_profit_cash,
        "tax_before_deduction": tax_before_deduction,
        "tax_deduction": tax_deduction,
        "tax_deduction_limit": deduction_limit,
        "tax_payable": tax_payable,
        "deductible_contrib": deductible_contrib,
        "contrib_self": contrib_self,
        "contrib_workers": contrib_workers,
        "contrib_self_extra": owner_extra,
        "contrib_self_extra_base": owner_extra_base,
        "owner_extra": owner_extra,
        "owner_extra_base": owner_extra_base,
        "fixed_contrib": data.fixed_contrib,
        "has_employees_patent_limit": 1 if has_employees_limit else 0,
        "total_payments": total_burden,
        "patent_cost_year": data.patent_cost_year,
        "patent_pvd_period": manual_pvd,
        "patent_pvd_used": pvd_used,
        "patent_pvd_auto": 0 if manual_pvd > 0 else 1,
        "patent_pvd_source": pvd_source,
    }

    return CalcResult(
        regime="patent",
        title="ПСН (патент)",
        revenue=revenue,
        expenses=expenses_total,
        tax=tax_payable,
        vat=0.0,
        insurance=contrib_self + contrib_workers,
        total_burden=total_burden,
        burden_percent=burden_percent,
        net_profit=net_profit_cash,
        extra=extra,
    )

File: static/js/explanations/builders/patent.js
(function (global) {
    const format = global.ExplanationsFormat;
    if (!format) {
        return;
    }
    const { fmtMoney, fmtPercent, renderMarkdown } = format;

    function buildPatentMarkdown(params) {
        const revenue = params.revenue || 0;
        const expenses = params.expensesWithoutVat || params.expenses || 0;
        const patentCost = params.patentCostYear !== undefined
            ? params.patentCostYear
            : (params.patentTaxBeforeDeduction !== undefined ? params.patentTaxBeforeDeduction : 0);
        const taxBeforeDeduction = params.patentTaxBeforeDeduction !== undefined
            ? params.patentTaxBeforeDeduction
            : patentCost;
        const fixedContrib = params.fixedContrib || 0;
        const contribWorkers = params.contribWorkers || params.baseInsurStd || 0;
        const ownerExtra = params.contribSelfExtra || params.ownerExtra || 0;
        const pvdUsed = params.patentPvdUsed || params.contribSelfExtraBase || params.ownerExtraBase || 0;
        const manualPvd = params.patentPvdPeriod || 0;
        const patentPvdSource = params.patentPvdSource || (params.patentPvdAuto ? 'auto' : '');
        const isAutoPvd = patentPvdSource === 'auto';
        const contribSelfTotal = fixedContrib + ownerExtra;
        const totalContrib = contribSelfTotal + contribWorkers;
        const netProfitCash = params.netProfitCash !== undefined ? params.netProfitCash : params.netProfit;
        const hasEmployeesLimit = (params.hasEmployeesPatentLimit || 0) > 0;
        const employeesCount = params.employeesCount || params.employees || 0;
        const payroll = params.payroll || params.baseFot || 0;
        const hasEmployees =
            hasEmployeesLimit
            || contribWorkers > 0
            || employeesCount > 0
            || payroll > 0;
        const limitPercent = hasEmployees ? 0.5 : 1;
        const limitPercentText = hasEmployees ? '50% (есть работники)' : '100% (нет работников)';
        const calculatedLimit = taxBeforeDeduction * limitPercent;
        const taxLimit = params.patentTaxDeductionLimit !== undefined ? params.patentTaxDeductionLimit : calculatedLimit;
        const patentReduction = Math.min(totalContrib, taxLimit);
        const taxDeduction = params.patentTaxDeduction !== undefined ? params.patentTaxDeduction : patentReduction;
        const taxPayableCalculated = Math.max(taxBeforeDeduction - taxDeduction, 0);
        const taxPayable = params.patentTaxPayable !== undefined
            ? params.patentTaxPayable
            : params.tax !== undefined
                ? params.tax
                : taxPayableCalculated;
        const baseForShare = taxBeforeDeduction || 0;
        const patentReducedShare = baseForShare > 0 ? patentReduction / baseForShare : 0;
        const patentRatePercent = 6;

        const limitFromCost = patentCost * limitPercent;
        const displayLimit = limitFromCost > 0 ? limitFromCost : taxLimit;
        const lines = [
            '# Пояснение к расчёту налоговой нагрузки (ПСН, патент)',
            '',
            '## 1. Общие параметры',
            `Выручка за год: ${fmtMoney(revenue)} ₽ (на величину патента не влияет, налог фиксирован).`,
            `Стоимость патента на год (ввод пользователя): ${fmtMoney(patentCost)} ₽.`,
            `Основные расходы (себестоимость + аренда + прочие + ФОТ): ${fmtMoney(expenses)} ₽ — для ПСН они не влияют на налог, но уменьшают чистую прибыль.`,
            '',
            '## 2. Страховые взносы',
            `Взносы за работников (30% от ФОТ): ${fmtMoney(contribWorkers)} ₽.`,
            isAutoPvd
                ? `ПВД для расчёта 1% берётся автоматически из стоимости патента: ${fmtMoney(patentCost)} ₽ / ${patentRatePercent}% = ${fmtMoney(pvdUsed)} ₽.`
                : `ПВД для расчёта 1% введён вручную: ${fmtMoney(manualPvd || pvdUsed)} ₽ (если не заполнить, он берётся из стоимости патента).`,
            `Фиксированный взнос ИП: ${fmtMoney(fixedContrib)} ₽.`,
            `1% свыше 300 000 ₽ от ПВД: max(${fmtMoney(pvdUsed)} ₽ − 300 000 ₽, 0) × 1% = ${fmtMoney(ownerExtra)} ₽.`,
            `Итого взносы ИП за себя (фиксированный + 1%): ${fmtMoney(contribSelfTotal)} ₽.`,
            '',
            '## 3. Общая сумма взносов',
            `Всего взносов, доступных для уменьшения патента: ${fmtMoney(totalContrib)} ₽ (работники + ИП).`,
            '',
            '## 4. Уменьшение стоимости патента',
            hasEmployees
                ? 'Есть сотрудники или выплачивался ФОТ, поэтому закон ограничивает уменьшение патента 50% его стоимости.'
                : 'Нет сотрудников и не выплачивался ФОТ, поэтому патент можно уменьшить на всю стоимость (100%).',
            limitFromCost > 0
                ? `Лимит уменьшения составляет ${limitPercentText}: ${fmtMoney(patentCost)} ₽ × ${hasEmployees ? '50%' : '100%'} = ${fmtMoney(limitFromCost)} ₽.`
                : `Лимит уменьшения задаётся правилом ${hasEmployees ? '50%' : '100%'} от стоимости патента.`,
            `Правило расчёта суммы уменьшения: берём меньшую величину из доступных взносов (${fmtMoney(totalContrib)} ₽) и рассчитанного лимита (${fmtMoney(displayLimit)} ₽).`,
            `Налог к уплате после уменьшения: ${fmtMoney(taxPayable)} ₽.`,
            '',
            '## 5. Итог',
            `Обязательные платежи (налог после уменьшения + взносы): ${fmtMoney(taxPayable)} ₽ + ${fmtMoney(totalContrib)} ₽ = ${fmtMoney(taxPayable + totalContrib)} ₽.`,
            `Чистая прибыль (cash): ${fmtMoney(revenue)} ₽ − ${fmtMoney(expenses)} ₽ − ${fmtMoney(taxPayable)} ₽ − ${fmtMoney(totalContrib)} ₽ = ${fmtMoney(netProfitCash)} ₽.`,
        ];

        return renderMarkdown(lines);
    }

    global.ExplanationBuilders = global.ExplanationBuilders || {};
    global.ExplanationBuilders.buildPatentMarkdown = buildPatentMarkdown;
})(window);
